JDK-8257588: Make os::_page_sizes a bitmask

tstuefe opened the pull request:
Hi,

may I please have reviews for the following very small improvement:

While discussing JDK-8243315 [1], and aiming to make planned changes like JDK-8256155 [2] easier:

```
size_t os::_page_sizes[os::page_sizes_max];
```

is an array used to keep all page sizes the hotspot can use. It is sorted by size and filled in at initialization time.

Coding dealing with this can be simplified by making this a set (which is very easy since all page sizes are power-2-values so they lend themselves nicely to a bitmap).

That has the following advantages:
- makes adding new sizes simple since we do not have to re-sort the array. Coding is easier to read too.
- it makes it possible to encode a set of page sizes in one number, so we can hand a set-of-page-sizes around as a value.

-----

The patch adds a new class, os::PagesizeSet, which is a bitmap containing page sizes. It adds gtests for this class. It replaces the old os::_page_sizes with an object of that class. It also removes an unused function.


Testing:
- nightlies at SAP
- manual tests with UseLargePages
- the new gtests
- gh actions (with x86 still failing because of unrelated issues)

Thank you,

Thomas

[1] https://bugs.openjdk.java.net/browse/JDK-8243315
[2] https://bugs.openjdk.java.net/browse/JDK-8256155

/label hotspot-dev

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8257588](https://bugs.openjdk.java.net/browse/JDK-8257588): Make os::_page_sizes a bitmask


### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to 919d90bb274e7bd55f5443d3943adaf5ef4926a2


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1522/head:pull/1522`
`$ git checkout pull/1522`

tschatzl commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -100,19 +100,30 @@ class os: AllStatic {
+    void add(size_t pagesize);
+    bool is_set(size_t pagesize) const;
+    // given a page size, return the next smaller page size in this set, or 0.
+    size_t next_smaller(size_t pagesize) const;
```
Please let all documentation sentences start with upper case.

tschatzl commented on src/hotspot/share/runtime/os.cpp:
```diff
@@ -1857,3 +1845,82 @@ void os::naked_sleep(jlong millis) {
+      st->print(SIZE_FORMAT "k", sz / K);
+    } else if (sz < G) {
+      st->print(SIZE_FORMAT "m", sz / M);
+    } else {
```
The prefixes for mega/gigabyte are "M" and "G" instead of "m" and "g" respectively

tschatzl commented on src/hotspot/share/runtime/os.cpp:
```diff
@@ -1857,3 +1845,82 @@ void os::naked_sleep(jlong millis) {
+  if (v2 == 0) {
+    return 0;
+  }
+  while ((v2 & 1) == 0) {
```
Maybe the `count_trailing_zeros()` method could be used here instead of doing the same manually.

tschatzl reviewed (changes requested):
To be really nitpicky: the subject of the PR should be made more specific to mention that the new set is a bitset not just a set. I.e. os::page_sizes has already previously been a set (of integers), so where's the change? :P

tstuefe commented on src/hotspot/share/runtime/os.cpp:
```diff
@@ -1857,3 +1845,82 @@ void os::naked_sleep(jlong millis) {
+
+size_t os::PagesizeSet::smallest() const {
+  assert(min_page_size() > 0, "Sanity");
+  return next_larger(min_page_size() / 2);
```
Not sure what you mean.

PagesizeSet::smallest() returns the smallest page in the page size set, which may be any size, or 0 if empty. 

os::_page_sizes should always contain os::vm_page_size(), but that too is not the same as os::min_page_size(). os::min_page_size() is just a constant, equal or smaller than the smallest page size across all platforms.

The naming is confusing, and probably it could be just a static const instead of a function. But I don't want to touch that with this change. But there are probably a number of misuses of os::min_page_size() where the author actually meant os::vm_page_size().

tschatzl replied:
Okay, apologies for falling into the mentioned trap.

tstuefe replied:
No problem. We should probably improve this.

tstuefe commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -100,19 +100,30 @@ class os: AllStatic {
 
  public:
-  enum { page_sizes_max = 9 }; // Size of _page_sizes array (8 plus a sentinel)
+
```
Would it be okay to leave this public anyway? It is used from the gtests. I could make a fixture class and make it friend of os but that would be similarly ugly.

tstuefe commented on src/hotspot/share/runtime/os.cpp:
```diff
@@ -1857,3 +1845,82 @@ void os::naked_sleep(jlong millis) {
+  _v |= pagesize;
+}
+
+// returns true if given page size is part of the set
```
Not possible, since its used from within os::Linux, os::Aix etc.

tschatzl replied:
Your comment is probably related to the comment of making the `PageSizeSet` class private. Hmm, I did try that out though and compiled locally on Linux.... maybe I messed up somewhere. Leave it public then.

Otherwise I am a bit confused how this statement relates to the request to move the code comment to the .hpp file.

tstuefe replied:
You know, communication was a bit easier when we wrote mails :) 
I thought you meant "make add() private". Sorry for my confusion.

tstuefe commented:
Hi Thomas,

thanks again for the review. I hope I addressed all points. I completely removed the comments from the cpp file and corrected them in the hpp file, which removes the need to keep them in sync. And I rather have them in the hpp file, that makes my IDE happy.

Cheers, Thomas

mgkwill commented:
+1 on the general thrust of this change. 

I'd be happy to incorporate into https://github.com/openjdk/jdk/pull/1153 & JDK-8256155 [2] if they are approved.

stefank reviewed (commented):
Just  a few drive-by comment: Could PagesizeSet be named PageSizeSet. The rest of the code separates the word, so I think that name would be better. The same for some of the parameters that are named pagesize.

Also, I think you could have called the class PageSizes. I don't think the fact that this is a 'set' matters for the usages of the class.  It also makes the variable name match the type name, which I guess is an indication that it wouldn't be such a bad name:
 ```
static PageSizes        _page_sizes;`
```
vs
```
  static PagesizeSet        _page_sizes;
```

stefank commented on src/hotspot/share/runtime/os.cpp:
```diff
@@ -1857,3 +1845,73 @@ void os::naked_sleep(jlong millis) {
+
+bool os::PagesizeSet::is_set(size_t pagesize) const {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
+  return (_v & pagesize) > 0;
```
Why is this `> 0` and not simply `!= 0`

stefank commented on src/hotspot/share/runtime/os.cpp:
```diff
@@ -1857,3 +1845,73 @@ void os::naked_sleep(jlong millis) {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
+  // mask out all pages sizes >= pagesize:
+  uintx v2 = _v & (pagesize - 1);
+  if (v2 > 0) {
+    return round_down_power_of_2(v2);
+  }
+  return 0;
```
In next_larger you use the for:
```
if (v2 == 0) {
  return 0;
}
return ...
```
I think it would be nice to be consistent between the two functions.

tstuefe replied:
> Just a few drive-by comment: Could PagesizeSet be named PageSizeSet. The rest of the code separates the word, so I think that name would be better. The same for some of the parameters that are named pagesize.
> 
> Also, I think you could have called the class PageSizes. I don't think the fact that this is a 'set' matters for the usages of the class. It also makes the variable name match the type name, which I guess is an indication that it wouldn't be such a bad name:
> 
> ```
> static PageSizes        _page_sizes;`
> ```
> 
> vs
> 
> ```
>   static PagesizeSet        _page_sizes;
> ```

Hi Stefan,

thanks for your review! I adapted most of your suggestions.

Cheers, Thomas

stefank commented on src/hotspot/share/runtime/os.cpp:
```diff
@@ -1857,3 +1845,73 @@ void os::naked_sleep(jlong millis) {
+}
+
+size_t os::PagesizeSet::next_smaller(size_t pagesize) const {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
```
You're using INTPR_FORMAT to print a size_t. I guess this is done to print hex values. We have a SIZE_FORMAT_HEX for that use case.

tstuefe replied:
Sure

stefank commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -100,19 +100,30 @@ class os: AllStatic {
+  public:
+    PagesizeSet() : _v(0) {}
+    void add(size_t pagesize);
+    bool is_set(size_t pagesize) const;
```
Could this be private. It's an implementation detail that you use a bit set and have "set" bits. If not, maybe consider a more high-level name. is_added, is_registered, contains?

mgkwill replied:
I'd prefer if this stays public, I've already found it useful, but I agree that the name could be more informative. Perhaps contains(size_t pagesize)?

tstuefe replied:
Cannot be private since its used from other nested classes (os::<platform>). But I can rename it to contains()

stefank commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -100,19 +100,30 @@ class os: AllStatic {
+
+  // A simple value class holding a set of page sizes (similar to sigset_t)
+  class PagesizeSet {
+    uintx _v;
```
All in-parameters are size_t. It seems like this could be size_t as well.

tstuefe replied:
This is my personal taste; I dislike keeping anything a size_t which is not a memory range size (e.g. I dislike its use as an array index). But it does not matter much to me here since size_t is guaranteed to be unsigned.

stefank commented on src/hotspot/share/runtime/os.cpp:
```diff
@@ -1857,3 +1845,72 @@ void os::naked_sleep(jlong millis) {
+}
+
+size_t os::PageSizes::smallest() const {
+  assert(min_page_size() > 0, "Sanity");
+  return next_larger(min_page_size() / 2);
```
I'm not sure why you use min_page_size() / 2 instead of another small enough power_of_2, say the smallest one:
```
return next_larger(1);
```
and thereby get rid of the dependency on the odd min_page_size().

tstuefe replied:
True. Earlier implementation had a lop in next_larger. But now it does not matter. 1 is fine.

tstuefe replied:
(Note that min_page_size should be really a constant, and named differently. I have this mentally marked as another small cleanup)

stefank commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -100,19 +100,30 @@ class os: AllStatic {
-    _page_sizes[0] = default_page_size;
-    _page_sizes[1] = 0; // sentinel
-  }
+  static PageSizes        _page_sizes;
```
Indentation is off by two spaces

stefank commented on test/hotspot/gtest/runtime/test_os.cpp:
```diff
@@ -587,3 +587,85 @@ TEST_VM(os, find_mapping_3) {
+static const int min_page_size_log2 = exact_log2(os::min_page_size());
+static const int max_page_size_log2 = (int)(sizeof(size_t) * 8);
+
+TEST_VM(os, pagesizeset_test_range) {
```
pagesizeset name needs to be updated.

stefank commented on test/hotspot/gtest/runtime/test_os.cpp:
```diff
@@ -587,3 +587,85 @@ TEST_VM(os, find_mapping_3) {
+static const int max_page_size_log2 = (int)(sizeof(size_t) * 8);
+
+TEST_VM(os, pagesizeset_test_range) {
+  for (int bit = min_page_size_log2; bit < max_page_size_log2; bit ++) {
```
bit ++ => bit++ : the same in the rest of the file.

stefank commented on test/hotspot/gtest/runtime/test_os.cpp:
```diff
@@ -587,3 +587,85 @@ TEST_VM(os, find_mapping_3) {
+}
+
+static const int min_page_size_log2 = exact_log2(os::min_page_size());
+static const int max_page_size_log2 = (int)(sizeof(size_t) * 8);
```
Could this be simply BitsPerWord?

stefank commented on test/hotspot/gtest/runtime/test_os.cpp:
```diff
@@ -587,3 +587,85 @@ TEST_VM(os, find_mapping_3) {
+    for (int bit2 = min_page_size_log2; bit2 < max_page_size_log2; bit2 ++) {
+      const size_t s =  (size_t)1 << bit;
+      const size_t s2 = (size_t)1 << bit2;
+      //tty->print_cr(SIZE_FORMAT " - " SIZE_FORMAT, s, s2);
```
Remove?

stefank commented on test/hotspot/gtest/runtime/test_os.cpp:
```diff
@@ -587,3 +587,85 @@ TEST_VM(os, find_mapping_3) {
+      const size_t s2 = (size_t)1 << bit2;
+      //tty->print_cr(SIZE_FORMAT " - " SIZE_FORMAT, s, s2);
+      os::PageSizes pss;
+      // Empty set
+      for (int bit3 = min_page_size_log2; bit3 < max_page_size_log2; bit3 ++) {
+        const size_t s3 = (size_t)1 << bit3;
+        ASSERT_FALSE(pss.contains(s3));
+      }
+      ASSERT_EQ((size_t)0, pss.smallest());
+      ASSERT_EQ((size_t)0, pss.largest());
```
This doesn't have to be inside a double nested for loop. This could be single check done once.

stefank commented on test/hotspot/gtest/runtime/test_os.cpp:
```diff
@@ -587,3 +587,85 @@ TEST_VM(os, find_mapping_3) {
+  char buffer[256];
+  stringStream ss(buffer, sizeof(buffer));
+  pss.print_on(&ss);
+  // tty->print_cr("%s", buffer);
```
Remove?

stefank reviewed (commented):
I think this mostly looks good, but there are a number of cleanups that I'd like you to consider.

tstuefe commented on src/hotspot/share/runtime/os.cpp:
```diff
@@ -1857,3 +1845,72 @@ void os::naked_sleep(jlong millis) {
+
+////// Implementation of PageSizes
+
+void os::PageSizes::add(size_t pagesize) {
```
Oh missed that in your first review, sorry.

tstuefe commented on src/hotspot/share/runtime/os.cpp:
```diff
@@ -1857,3 +1845,72 @@ void os::naked_sleep(jlong millis) {
+  if (pagesize == max_power_of_2<size_t>()) { // Shift by 32/64 would be UB
+    return 0;
+  }
+  int l = exact_log2(pagesize) + 1;
```
I like yours more. Its clearer.

tstuefe commented:
Hi Stefan,

I pushed a new commit with the changes you requested.

Thanks, Thomas

stefank reviewed (approved):
Thanks! Looks good.

tstuefe commented:
> Thanks! Looks good.

Thank you Stefan.

tstuefe commented:
Thanks Thomas

tstuefe commented:
/integrate

mgkwill commented on test/hotspot/gtest/runtime/test_os.cpp:
```diff
@@ -85,29 +85,29 @@ TEST_VM(os, page_size_for_region_alignment) {
+    for (size_t s = os::page_sizes().largest(); s != 0; s = os::page_sizes().next_smaller(s)) {
+      const size_t expected = os::page_sizes().next_smaller(s);
+      if (expected != 0) {
+        size_t actual = os::page_size_for_region_unaligned(expected - 17, 1);
```
@tstuefe should **expected** here be **s** ?

It seems like we are trying to compare the next smaller page size of S, with a slightly smaller size of the size S.

`os::page_size_for_region_unaligned(expected - 17, 1);`
vs.
`os::page_size_for_region_unaligned(s- 17, 1);`

Running these tests with 2 largepage sizes (3 total sizes) fails, however default of 2 page_sizes passes (4k and 2m or 1g):
(./hotspot/variant-server/libjvm/gtest/gtestLauncher -jdk:./images/jdk -Xms2G -Xmx2G -XX:+UseLargePages -XX:LargePageSizeInBytes=2M)
> [----------] 17 tests from os                                                                                                                                                                          [ RUN      ] os.page_size_for_region_vm                                                                                                                                                                [       OK ] os.page_size_for_region_vm (0 ms)                                                                                                                                                         [ RUN      ] os.page_size_for_region_aligned_vm                                                                                                                                                        [       OK ] os.page_size_for_region_aligned_vm (0 ms)                                                                                                                                                 [ RUN      ] os.page_size_for_region_alignment_vm                                                                                                                                                      [       OK ] os.page_size_for_region_alignment_vm (0 ms)                                                                                                                                               [ RUN      ] os.page_size_for_region_unaligned_vm                                                                                                                                                      test/hotspot/gtest/runtime/test_os.cpp:106: Failure                                                                                                                                                    Expected equality of these values:                                                                                                                                                                       actual                                                                                                                                                                                                   Which is: 4096                                                                                                                                                                                       expected                                                                                                                                                                                                 Which is: 2097152                                                                                                                                                                                  [  FAILED  ] os.page_size_for_region_unaligned_vm (0 ms)

mgkwill replied:
See https://github.com/openjdk/jdk/pull/1719 for fix.

tstuefe replied:
Yes, you are right, it should. I filed JDK-8257989. Feel free to fix it.

Thanks, Thomas

tstuefe replied:
Oh, I see you already have a PR open, I reply there.

mgkwill replied:
Fixed in #1719. Thanks Thomas.
